cmplxs/cmplxs.go

```diff
@@ -0,0 +1,954 @@
+}
+
+func ulpDiff(a, b uint64) uint64 {
+	if a > b {
+		return a - b
+	}
+	return b - a
+}
+
+// EqualLengths returns true if all of the slices have equal length,
+// and false otherwise. Returns true if there are no input slices.
+func EqualLengths(slices ...[]complex128) bool {
+	// This length check is needed: http://play.golang.org/p/sdty6YiLhM
+	if len(slices) == 0 {
+		return true
+	}
+	l := len(slices[0])
+	for i := 1; i < len(slices); i++ {
+		if len(slices[i]) != l {
+			return false
+		}
+	}
+	return true
+}
+
+// Find applies f to every element of s and returns the indices of the first
+// k elements for which the f returns true, or all such elements
+// if k < 0.
+// Find will reslice inds to have 0 length, and will append
+// found indices to inds.
+// If k > 0 and there are fewer than k elements in s satisfying f,
+// all of the found elements will be returned along with an error.
+// At the return of the function, the input inds will be in an undetermined state.
+func Find(inds []int, f func(complex128) bool, s []complex128, k int) ([]int, error) {
+	// inds is also returned to allow for calling with nil
+
+	// Reslice inds to have zero length
+	inds = inds[:0]
+
+	// If zero elements requested, can just return
+	if k == 0 {
+		return inds, nil
+	}
+
+	// If k < 0, return all of the found indices
+	if k < 0 {
+		for i, val := range s {
+			if f(val) {
+				inds = append(inds, i)
+			}
+		}
+		return inds, nil
+	}
+
+	// Otherwise, find the first k elements
+	nFound := 0
+	for i, val := range s {
+		if f(val) {
+			inds = append(inds, i)
+			nFound++
+			if nFound == k {
+				return inds, nil
+			}
+		}
+	}
+	// Finished iterating over the loop, which means k elements were not found
+	return inds, errors.New("cmplxs.Find: insufficient elements found")
+}
+
+// HasNaN returns true if the slice s has any values that are NaN and false
+// otherwise.
+func HasNaN(s []complex128) bool {
+	for _, v := range s {
+		if cmplx.IsNaN(v) {
+			return true
+		}
+	}
+	return false
+}
+
+// Hermitian generates the hermitian (conjugate transpose) of s with r rows and c cols,
+// and stores in s.
+// It panics if the length of s is not equal to r * c.
+func Hermitian(r, c int, s []complex128) {
+	if (r * c) != len(s) {
+		panic("cmplxs.Hermitian: rows & cols do not match with length of s")
+	}
+	dst := make([]complex128, len(s))
+	for i := 0; i < r; i++ {
+		for j := 0; j < c; j++ {
+			dst[i+r*j] = cmplx.Conj(s[i*c+j])
+		}
+	}
+	copy(s, dst)
+}
+
+// HermitianTo generates the hermitian (conjugate transpose) of s with r rows and c cols,
+// and stores in dst.
+// It panics if the length of s is not equal to r * c and if the length of s and dst are
+// not equal.
+func HermitianTo(r, c int, dst, s []complex128) []complex128 {
+	if (r * c) != len(s) {
+		panic("cmplxs.HermitianTo: rows & cols do not match with length of s")
+	}
+	if len(s) != len(dst) {
+		panic("cmplxs.HermitianTo: length of s and dst do not match")
+	}
```

Comment:
Similar comments here: no parens needed, prefix with cmplxs, no not use '&' as an abbreviation for "and".